clang-uml: new Portfile

bkryza opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 15.3.2 24D81 arm64
Xcode 16.2 16C5032a

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

bkryza commented:
Added patch to fix pthreds and libunwind detection in MacPorts.

bkryza commented:
@reneeotten Thank you for reviewing the PR and the comments. 

To answer your question, yes `clang-uml` needs to use and be compiled with `macports-clang-*` due to the fact that it uses low level LLVM/Clang API's to analyze the Abstract Syntax Tree generated by Clang after parsing a C++ project so that it can generate various types of diagrams from the code (class diagrams, sequence diagrams, etc.). In this respect it is very similar to [include-what-you-use](https://github.com/macports/macports-ports/blob/dc67d6aacdcfaaa7f248173492ff2ada060e297d/devel/include-what-you-use/Portfile) (I've initially based my Portfile on this one). With Apple Clang, there might be various changes or incompatibilities in those API's, hopefully with these changes only `macports-clang-*` compilers will be used now.

I tried to address all of your remarks, in addition I made the following changes:
- removed `platforms` and `supported_archs` as it was giving me a warning after upgrading my ports today (can't remember why I added these in the first place)
- instead of setting a fixed llvm version, I added `compiler.whitelist` to 3 latest Clang versions in case someone doesn't want to upgrade to 21 or even 20 (I tested it locally and it works).

I haven't squashed the commits in the PR yet so you can see the changes I made today.

bkryza commented on devel/clang-uml/Portfile:
```diff
@@ -0,0 +1,59 @@
+PortGroup           cmake 1.1
+
+name                clang-uml
+version             0.6.1
```
Done.

bkryza commented on devel/clang-uml/Portfile:
```diff
@@ -0,0 +1,59 @@
+    xinstall -m 755 -s ${workpath}/build/src/clang-uml ${destroot}${prefix}/bin
+    xinstall -m 755 -d ${destroot}${prefix}/share/doc/${name}
+    xinstall -m 644 -W ${worksrcpath} README.md LICENSE.md CHANGELOG.md AUTHORS.md \
+        ${destroot}${prefix}/share/doc/${name}
```
Done.

reneeotten commented:
> With Apple Clang, there might be various changes or incompatibilities in those API's, hopefully with these changes only `macports-clang-*` compilers will be used now.

>     * instead of setting a fixed llvm version, I added `compiler.whitelist` to 3 latest Clang versions in case someone doesn't want to upgrade to 21 or even 20 (I tested it locally and it works).

thanks for the updates @bkryza ! I think the compiler selection still doesn't work in the way you want/intended... In theory, the end-user can now select a compiler by adding `configure.compiler` to the command line when installing the port. But that's not very user-friendly and typically not what we do; ideally if you want to allow different compiler, this should be done using "variants".

If you just want to pick a version (be it the latest `21` or for now still `20` if you think that's more stable) that is perfectly fine. However, don't you want to make sure that it will then also use `/opt/local/libexec/llvm-XY/bin/clang++` when compiling the package. As you'll see in the logs once the CI is finished, I think that's not what is happening right now. So that's probably something that you'll want to fix, perhaps by using the `compilers` PortGroup and setting `macports-clang` compilers (I guess you technically could allow all of them that support CXX17, not necessarily only the last three?). As I said, I am not that experienced with it, but you could look at some other Portfiles that use it or read through the descriptions in the [PortGroup](https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/compilers-1.0.tcl). A Wiki page regarding compiler selection in MacPorts is [here](https://trac.macports.org/wiki/CompilerSelection).

bkryza commented:
@reneeotten Thanks, I looked into the variants and port groups but I think at this point it would be a bit of an overkill for this tool. The users can still use `clang-uml` compiled and linked with a specific version of LLVM on C++ projects build with any other compiler (other LLVM versions, GCC, XCode), assuming that LLVM version supports all language features in that project. 

So for now I fixed both Clang compiler and Clang library in the `Portfile` - if I understood the `configure.compiler` correctly - to LLVM 20 (current stable version of `clang-uml` 0.6.2 was released in June, so it wasn't tested with the latest 21.1.1), I'll try to update it with the next `clang-uml` release.

I also cleaned up the commit history.

bkryza commented:
@reneeotten Thank you!
